rdtools/soiling.py

```diff
@@ -1052,3 +1065,1532 @@ def monthly_soiling_rates(soiling_interval_summary, min_interval_length=14,
+        if soiling_signal_strength < soiling_significance_knob:
+            if verbose:
+                print('Soiling signal is small relative to the noise')
+            small_soiling_signal = True
+            df_out.SR_high = 1.0
+            df_out.SR_low = 1.0 - SR_amp
+
+        return df_out, degradation, soiling_loss, residual_shift, RMSE, \
+            small_soiling_signal, adf_res
+
+    def run_bootstrap(self,
+                      reps=512,
+                      confidence_level=68.2,
+                      verbose=False,
+                      degradation_method='YoY',
+                      process_noise=1e-4,
+                      order_alternatives=(('SR', 'SC', 'Rd'),
+                                          ('SC', 'SR', 'Rd')),
+                      detection_tuner_alternatives=(.25, .75),
+                      pruning_tuner_alternatives=(1/1.5, 1.5),
+                      forward_fill_alternatives=(True, False),
+                      **kwargs):
+        '''
+        Bootstrapping of CODS alforithm for uncertainty analysis, inherently accounting
+        for model and parameter choices.
+
+        First, calls on :py:func:`iterative_signal_decomposition` to fit N different
+        models. Next, bootstrap samples are generated based on these N initial model fits.
+        Each bootstrap sample is generated by bootstrapping the residuals of
+        the respective model fit, using circular block
+        bootstrapping, then multiplying these new residuals back onto the
+        model. Then, for each bootstrap sample, model parameters are randomly
+        chosen and the CODS model is fit to the bootstrapped signal.
+        The seasonal component is perturbed randomly and
+        divided out, so as to capture its uncertainty. In the end,
+        confidence intervals are calulated based on the percentile levels of the
+        collection of bootstrapped model fits.
+        The returned soiling ratio and rates are based on
+        the best fit of the initial N models. See [1]_ for more details.
+
+        Parameters
+        ----------
+        reps : int, default 512,
+            Number of bootstrap realizations to be run
+            minimum N, where N is the possible combinations of model
+            knobs/parameters defined in knob_alternatives
+        confidence_level : float, default 68.2
+            The size of the confidence intervals to return, in percent
+        verbose : bool, default False
+            Wheter or not to print information about progress
+        degradation_method : string, default 'YoY'
+            Either 'YoY' or 'STL'. If anything else, 'YoY' will be assumed.
+            Decides whether to use the YoY method [3] for estimating the
+            degradation trend (assumes linear trend), or the STL-method (does
+            not assume linear trend). The latter is slower.
+        order_alternatives : tuple of tuples, default (('SR', 'SC', 'Rd'), \
+            ('SC', 'SR', 'Rd'))
```

Comment:
This renders strangely -- I think this would fit all on one line without angering the linter (see https://github.com/NREL/rdtools/blob/master/.flake8), so probably that's the easiest solution.